### PR TITLE
normalization + dumping beam distribution at startdump

### PR DESCRIPTION
**Pull request by mfittere**
So this is not finished yet. I am kidn of in the middle ...

### CODE REVIEW

**Comment by kyrsjo** [comment]
Thanks for adding the PR; it's easier to follow when there is a PR.

Just to clarify, this is targeting finishing the issue #301 , and solving #303 and #299 right?

**Comment by kyrsjo** [comment]
Also you may want to merge from master to get the mods we did with NAFF...

File: SixTrack/sixtrack.s [hunk 1]
```diff
@@ -1190,13 +1190,15 @@
...
-                                                 !  (nomalisation of phase space)
-                                                 !  First index = -1 -> StartDUMP, filled differently;
-                                                 !  First index = 0  -> Unused.
-      double precision :: dump_clo (-1:nblz,6)   ! closed orbit used for FMA
+      double precision :: dumptas (-1:nblz,6,6) ! tas matrix used for FMA analysis
+                                                !  (nomalisation of phase space)
+                                                !  First index = -1 -> StartDUMP, filled differently;
+                                                !  First index = 0  -> Unused.
+      double precision :: dumptasinv (-1:nblz,6,6) ! inverse matrix of dumptas
+      double precision :: dumptasaux(6,6),dumptasinvaux(6,6) ! auxiliary variables
```
Comment (kyrsjo) [inline SixTrack/sixtrack.s]: Remove the aux stuff

**Comment by kyrsjo** [comment]
OK, I think I've figured out the mystery: It seems that the NORM_DUMP files are now only filled (except for the header, which is always filled) when fma_norm_flag == 1. The file is re-opened, truncated, and re-written for every single FMA that happens to the DUMP.

Even worse, the copying from to xyzv from xyzvdummy only happens if fma_norm_flag == 1, giving PLATO random memory garbage (probably leftovers from the previous FMA?) the chew on.

*fixing*

**Comment by kyrsjo** [comment]
OK, now it passes the test. Time to break it! :D

**Comment by kyrsjo** [comment]
Tests broken and reset. Bon weekend!

**Comment by kyrsjo** [comment]
Remember:
 - [ ] Update tests for NAFF canonicals
 - [ ] Add / modify Sixin.zip where needed
 - [ ] Add singletrackfile where needed

**Comment by kyrsjo** [comment]
Merged master - which contained NAFF - into this branch; merge went cleanly.

Also checked STARTdump for format 7 vs. the normalized version of format 3 provided by FMA, and it's a match (up to 513 turns):
```
 dump7_fma $ diff  START_DUMP_1 ../fma_binary/NORM_START_DUMP_1 | head -n 50
1,3c1
< # DUMP format #7, START=ip3             , number of particles=          60, dump period=           1, first turn=           1, last turn=         512, HIGH=F, FRONT=F
< # ID turn s[m] nx[1.e-3 sqrt(m)] npx[1.e-3 sqrt(m)] ny[1.e-3 sqrt(m)] npy[1.e-3 sqrt(m)] nsig[1.e-3 sqrt(m)] ndp/p[1.e-3 sqrt(m)] ktrack
< # closed orbit [mm,mrad,mm,mrad,1]  -1.956795435E-03 -3.393444660E-05 -2.206698166E-04 -5.365064302E-06  4.922952265E-03  4.201488863E-06
---
> # closorb  -1.956795435E-03 -3.393444660E-05 -2.206698166E-04 -5.365064302E-06  4.922952265E-03  4.201488863E-06
5a4
> # id turn pos[m] nx[1.e-3 sqrt(m)] npx[1.e-3 sqrt(m)] ny[1.e-3 sqrt(m)] npy[1.e-3 sqrt(m)] nsig[1.e-3 sqrt(m)] ndp/p[1.e-3 sqrt(m)] kt
30725a30725,245764
>         1      513      0.00000 -2.234881831E-03  4.341396928E-02  3.270556869E-03  1.126130145E-02 -1.335162822E-01  2.269629752E-01        0
>         2      513      0.00000 -2.235072024E-03  4.341405172E-02  3.270524757E-03  1.126131996E-02 -1.335162831E-01  2.269629736E-01        0
>         3      513      0.00000 -1.976231940E-03  4.493183560E-02  3.120126301E-03  1.171913357E-02 -1.335147673E-01  2.269634589E-01        0
>         4      513      0.00000 -1.976419881E-03  4.493191961E-02  3.120092626E-03  1.171915151E-02 -1.335147650E-01  2.269634595E-01        0
>         5      513      0.00000 -1.683063214E-03  4.644893516E-02  2.940686330E-03  1.217667371E-02 -1.335131748E-01  2.269639589E-01        0
>         6      513      0.00000 -1.683248807E-03  4.644902078E-02  2.940651045E-03  1.217669101E-02 -1.335131733E-01  2.269639597E-01        0
>         7      513      0.00000 -1.354215795E-03  4.796477798E-02  2.731137671E-03  1.263264044E-02 -1.335115063E-01  2.269644768E-01        0
>         8      513      0.00000 -1.354398954E-03  4.796486522E-02  2.731100727E-03  1.263265702E-02 -1.335115051E-01  2.269644795E-01        0
>         9      513      0.00000 -9.885379025E-04  4.947883060E-02  2.490416869E-03  1.308562172E-02 -1.335097581E-01  2.269650114E-01        0
>        10      513      0.00000 -9.887185342E-04  4.947891949E-02  2.490378227E-03  1.308563747E-02 -1.335097550E-01  2.269650102E-01        0
>        11      513      0.00000 -5.848874159E-04  5.099051506E-02  2.217505063E-03  1.353407151E-02 -1.335079255E-01  2.269655586E-01        0
>        12      513      0.00000 -5.850654127E-04  5.099060560E-02  2.217464686E-03  1.353408632E-02 -1.335079257E-01  2.269655619E-01        0
>        13      513      0.00000 -1.421335659E-04  5.249920789E-02  1.911437401E-03  1.397630685E-02 -1.335060046E-01  2.269661261E-01        0
>        14      513      0.00000 -1.423088323E-04  5.249930010E-02  1.911395262E-03  1.397632062E-02 -1.335060069E-01  2.269661245E-01        0
>        15      513      0.00000  3.408411484E-04  5.400423915E-02  1.571313156E-03  1.441050564E-02 -1.335040010E-01  2.269667064E-01        0
>        16      513      0.00000  3.406686841E-04  5.400433304E-02  1.571269221E-03  1.441051824E-02 -1.335040009E-01  2.269667027E-01        0
>        17      513      0.00000  8.651389198E-04  5.550489147E-02  1.196306493E-03  1.483470492E-02 -1.335018999E-01  2.269673000E-01        0
>        18      513      0.00000  8.649693784E-04  5.550498704E-02  1.196260745E-03  1.483471624E-02 -1.335019012E-01  2.269672995E-01        0
```

I temporarily disabled the NoNormDUMP for `fma_binary` to do this check.

**Comment by kyrsjo** [comment]
All dump and fma tests now pass with both CR and not-CR.

**Comment by kyrsjo** [comment]
OK, unless there are cases where the matrices are not valid I think we can bump the version number and merge. I'm running the "medium" tests now for standard SixTrack, CR, STF, and BOINC versions.

If there are such cases, such as 4D tracking or missing 6D linear optics block, this should be documented and tested for at runntime, both for FMA and DUMP as needed.

**Comment by mfittere** [comment]
yes, let's put a check on the 4d matrices and the missing line block. If the matrices are not set, all elements are 0, so it should be an easy check.

**Comment by kyrsjo** [comment]
The easiest for the 6D line block is probably to detect it already when parsing the input?

For 4D, look at how it is detected in fma_postpr.

**Comment by kyrsjo** [comment]
I noticed that we're writing some NaNs and other junk to fma_sixtrack when NOT doing the normalization and no matrix is avalable (i.e. no LINE block or LINE block not valid because of 4D); however this is not really a new problem - see #326 .

In general, I think once the docs are fully up2date I think we can declare success, bump the version, merge, and move forward.

**Pull request merged**